tsl/src/compression/batch_metadata_builder_bloom1.c

```diff
@@ -0,0 +1,601 @@
+/*
+ * This file and its contents are licensed under the Timescale License.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-TIMESCALE for a copy of the license.
+ */
+#include "postgres.h"
+
+#include <access/detoast.h>
+#include <catalog/pg_collation_d.h>
+#include <common/hashfn.h>
+#include <funcapi.h>
+#include <utils/builtins.h>
+#include <utils/typcache.h>
+
+#include <math.h>
+
+#include "compression.h"
+
+#include "batch_metadata_builder.h"
+
+#include "arrow_c_data_interface.h"
+
+#include "sparse_index_bloom1.h"
+
+#ifdef TS_USE_UMASH
+#include "import/umash.h"
+#endif
+
+/*
+ * There is a tradeoff between making the bloom filters selective enough but
+ * not too big. Testing on some real words datasets, the optimal value seems to
+ * be about 2% false positives. We also want to be able to reduce the filter to
+ * up to 64 bits, so that it fits in the main table. Hence the optimal number of
+ * hashes. This number actually gives a slightly different false positive rate,
+ * so this is what we ultimately use.
+ * Calculator: https://hur.st/bloomfilter/?p=0.02&m=64
+ */
+
+//#define BLOOM1_FALSE_POSITIVES 0.05
+//#define BLOOM1_HASHES 4
+
+#define BLOOM1_FALSE_POSITIVES 0.022
+#define BLOOM1_HASHES 6
+
+//#define BLOOM1_FALSE_POSITIVES 0.001
+//#define BLOOM1_HASHES 8
+
+/*
+ * Limit the bits belonging to the particular elements to a small contiguous
+ * region. This improves memory locality when building the bloom filter.
+ */
```

Comment:
should we expose this as GUC or make otherwise configurable

Reply:
Let's do this when we have configuration support for these indexes. The main issue is the number of hashes, I'm undecided whether I should add this to the metadata in the actual bloom1 values. It is also required for reading, and it should be chosen based on the false positive rate.